// react-native-firebase/dist/modules/firestore/CollectionReference.js
/**
 * 
 * CollectionReference representation wrapper
 */
import DocumentReference from './DocumentReference';
import Query from './Query';
import { firestoreAutoId } from '../../utils';

/**
 * @class CollectionReference
 */
export default class CollectionReference {

  constructor(firestore, collectionPath) {
    this._collectionPath = collectionPath;
    this._firestore = firestore;
    this._query = new Query(firestore, collectionPath);
  }

  get firestore() {
    return this._firestore;
  }

  get id() {
    return this._collectionPath.id;
  }

  get parent() {
    const parentPath = this._collectionPath.parent();
    return parentPath ? new DocumentReference(this._firestore, parentPath) : null;
  }

  add(data) {
    const documentRef = this.doc();
    return documentRef.set(data).then(() => Promise.resolve(documentRef));
  }

  doc(documentPath) {
    const newPath = documentPath || firestoreAutoId();

    const path = this._collectionPath.child(newPath);
    if (!path.isDocument) {
      throw new Error('Argument "documentPath" must point to a document.');
    }

    return new DocumentReference(this._firestore, path);
  }

  // From Query
  endAt(fieldValues) {
    return this._query.endAt(fieldValues);
  }

  endBefore(fieldValues) {
    return this._query.endBefore(fieldValues);
  }

  get() {
    return this._query.get();
  }

  limit(n) {
    return this._query.limit(n);
  }

  onSnapshot(onNext, onError) {
    return this._query.onSnapshot(onNext, onError);
  }

  orderBy(fieldPath, directionStr) {
    return this._query.orderBy(fieldPath, directionStr);
  }

  startAfter(fieldValues) {
    return this._query.startAfter(fieldValues);
  }

  startAt(fieldValues) {
    return this._query.startAt(fieldValues);
  }

  where(fieldPath, opStr, value) {
    return this._query.where(fieldPath, opStr, value);
  }
}
